### PR TITLE
object: print slow request id

### LINKED ISSUE #3767: Record the requestid of the object storage request in the log

### PR DESCRIPTION
**Pull request by zhijian-pro**
close #3767
- [x] azure
- [x] cos
- [x] ibmcos
- [x] ks3
- [x] obs
- [x] oos
- [x] qingstor
- [x] s3
- [x] tos

### CODE REVIEW

File: pkg/chunk/cached_store.go [hunk 1]
```diff
@@ -163,7 +163,7 @@ func (s *rSlice) ReadAt(ctx context.Context, page *Page, off int) (n int, err er
 		used := time.Since(st)
 		logger.Debugf("GET %s RANGE(%d,%d) (%s, %.3fs)", key, boff, len(p), err, used.Seconds())
```
Comment (davies) [inline pkg/chunk/cached_store.go:164]: add the request id here also

File: pkg/object/object_storage.go [hunk 2]
```diff
@@ -271,3 +271,47 @@ func ListAllWithDelimiter(store ObjectStorage, prefix, start, end string) (<-cha
...
 	return listed, nil
 }
+
+func init() {
+	ReqIDCache = reqIDCache{cache: make(map[string]reqItem)}
+	go ReqIDCache.Clean()
+}
+
+type reqItem struct {
+	ReqID string
```
Comment (davies) [inline pkg/object/object_storage.go:281]: lower case

File: pkg/object/object_storage.go [hunk 3]
```diff
@@ -271,3 +271,47 @@ func ListAllWithDelimiter(store ObjectStorage, prefix, start, end string) (<-cha
...
+}
+
+var ReqIDCache reqIDCache
+
+type reqIDCache struct {
+	sync.Mutex
+	cache map[string]reqItem
+}
+
+func (*reqIDCache) Clean() {
```
Comment (davies) [inline pkg/object/object_storage.go:292]: lower case

File: pkg/object/object_storage.go [hunk 4]
```diff
@@ -271,3 +271,47 @@ func ListAllWithDelimiter(store ObjectStorage, prefix, start, end string) (<-cha
...
+		for k, v := range ReqIDCache.cache {
+			if time.Since(v.time) > 2*time.Second {
+				delete(ReqIDCache.cache, k)
+			}
+		}
+		ReqIDCache.Unlock()
+	}
+}
+
+func (*reqIDCache) Put(key, reqID string) {
```
Comment (davies) [inline pkg/object/object_storage.go:304]: lower case

File: pkg/object/object_storage.go [hunk 5]
```diff
@@ -271,3 +271,47 @@ func ListAllWithDelimiter(store ObjectStorage, prefix, start, end string) (<-cha
...
+
+var ReqIDCache reqIDCache
+
+type reqIDCache struct {
+	sync.Mutex
+	cache map[string]reqItem
+}
+
+func (*reqIDCache) Clean() {
+	for range time.Tick(5 * time.Second) {
```
Comment (davies) [inline pkg/object/object_storage.go:293]: 1

File: pkg/object/object_storage.go [hunk 6]
```diff
@@ -271,3 +271,47 @@ func ListAllWithDelimiter(store ObjectStorage, prefix, start, end string) (<-cha
...
+		ReqIDCache.Lock()
+		for k, v := range ReqIDCache.cache {
+			if time.Since(v.time) > 2*time.Second {
```
Comment (davies) [inline pkg/object/object_storage.go:296]: 1

File: pkg/chunk/cached_store.go [hunk 7]
```diff
@@ -609,6 +601,20 @@ type cachedStore struct {
 	stageBlockDelay     prometheus.Counter
 }
 
+func logRequest(typeStr string, key string, param string, err error, used time.Duration) {
+	var info string
+	if object.ReqIDCache.Get(key) != "" {
```
Comment (davies) [inline pkg/chunk/cached_store.go:606]: if id := object.ReqIDCache.Get(key); id != "" {

File: pkg/chunk/cached_store.go [hunk 8]
```diff
@@ -609,6 +601,20 @@ type cachedStore struct {
...
+		info += fmt.Sprintf("RequestID: %s ", object.ReqIDCache.Get(key))
+	}
+	if err != nil {
+		info += err.Error()
+	}
+	logger.Debugf("slow request: %s %s %s(%v, %.3fs)", typeStr, key, param, info, used.Seconds())
```
Comment (davies) [inline pkg/chunk/cached_store.go:612]: this is not slow quest

File: pkg/chunk/cached_store.go [hunk 9]
```diff
@@ -609,6 +601,20 @@ type cachedStore struct {
...
+	if used > SlowRequest && logger.GetLevel() < logrus.DebugLevel {
```
Comment (davies) [inline pkg/chunk/cached_store.go:613]: don't need to check level

File: pkg/object/azure.go [hunk 10]
```diff
@@ -99,7 +103,10 @@ func (b *wasb) Put(key string, data io.Reader) error {
 	if b.sc != "" {
 		options.AccessTier = str2Tier(b.sc)
 	}
-	_, err := b.azblobCli.UploadStream(ctx, b.cName, key, data, &options)
+	resp, err := b.azblobCli.UploadStream(ctx, b.cName, key, data, &options)
+	if resp.RequestID != nil {
```
Comment (davies) [inline pkg/object/azure.go:107]: resp could be nil

Comment (zhijian-pro) [reply pkg/object/azure.go:107]: image
resp is not a pointer

**Review by davies — APPROVED** [review]

**Pull request merged**